### ISSUE TITLE
Ctrl-Click on new-style UNL may not navigate to target in other outline

### ISSUE THREAD

**Issue by tbpassin** [leo-editor/leo-editor#3670]
It looks like the construction of new-style UNLs makes assumptions about the location of a node's outline.  These assumptions can prevent Ctrl-Click navigation.

Example:
Here is the UNL of a node in LeoPyRef:

`unl:gnx://LeoPyRef.leo#ekr.20140901062324.18565`

If I paste this UNL into a node in my workbook, an attempt to Ctrl-Click on it fails with this error message:

`Not found: 'unl:gnx://LeoPyRef.leo#ekr.20140901062324.18565'`

The same failure occurs with the new version of old-style UNLs.

I have been under the impression that the new UNLs would support navigation to nodes in other outlines. Perhaps this will only work if the two outlines are in the same file tree.

**Comment by edreamleo** [comment]
@tbpassin Control-clicking unls searches all *open* outlines.

This may have implications for your zettlekasten work.

**Comment by tbpassin** [comment]
> @tbpassin Control-clicking unls searches all open outlines.

That is incorrect.  I just tried it again. I pasted a node's UNL into a node in the workbook.  Ctrl-clicking on the UNL string in the workbook failed to navigate to the original node even though its outline was open:

`Not found: 'unl:gnx://os-notes.leo#tom.20231124223010.1'`

**Comment by tbpassin** [comment]
> This may have implications for your zettlekasten work.

It uses gnx's and only within an outline.  I suppose it would have implications if I wanted to reference a node in a different outline, but that would be an ordinary reference - not part of the zettelkasten system.

I liked the old way where Leo would open the right outline if it weren't already open.

**Comment by edreamleo** [comment]
@tbpassin I am also going to delegate this issue to you. Please feel free to ask for help.

**Comment by tbpassin** [comment]
> @tbpassin Control-clicking unls searches all _open_ outlines.

That was what was supposed to happen but there were a few errors that prevented it.  I sorted that out, added a search of the files in the Recent Files list,  and added a few other files one might like to link to, like LeoPyRef.leo and LeoDocs.leo, so that they can be opened and navigated to even if they aren't in the recent files list.

**Comment by boltex** [comment]
@tbpassin @edreamleo I've now made UNL work in LeoJS (and soon in LeoInteg after i add a small modification to the server) :)

I've also made them open the leo file if not already opened successfully in LeoJS !  ( Having Leo automatically open an outline and navigate to a specific node from an UNL link is pretty cool feature! So I hope you can make it also work in Leo 😄  )

But, I've had trouble with the provided regex called **unl_regex** in **leoGlobals.py** to locate UNLs so I've made my own as I describe in the other Leo issue about UNLs here : https://github.com/leo-editor/leo-editor/issues/3667

Hope this helps!

**Comment by tbpassin** [comment]
Isn't it the case that JS regexes are a little different from Python ones?  Maybe that's behind part of what you have found.

**Comment by boltex** [comment]
@tbpassin Yes. But I've taken this into account.

**Comment by boltex** [comment]
@tbpassin In the specific case of this issue, I made some tests and I can confirm that UNL without the full path (that is only the file name such as unl:gnx://LeoPyRef.leo#ekr.20140901062324.18565 ) do not work even if the file is located in the same path as the current opened commander outline. (it only works in the same outline. )

Also, in another issue, https://github.com/leo-editor/leo-editor/issues/3667#issuecomment-1871685050 I've notified @edreamleo  that UNLs  without the 'outline' part
e.g. unl:gnx://#ekr.20140901062324.18565
should work in the currently opened Commander, but they do not.

Maybe this is related because it's a somewhat very similar issue.

**Comment by boltex** [comment]
@edreamleo @tbpassin 
To help visualize, After trying them out extensively, here is a comprehensive list of all possible UNL types and their effect when ctrl-clicked in Leo. (Please correct me if I'm wrong in the assumptions written in bold above each class of UNLs.)

**outline full path (work well even with backslashes and get opened if not opened)**
```
unl:gnx://C:\prog\testmarkdown\testmd1.leo#felix.20231224140524.1
unl:gnx://C:/prog/testmarkdown/testmd1.leo#felix.20231224140524.1

unl://C:/prog/testmarkdown/testmd1.leo#@button test2
unl://C:\prog\testmarkdown\testmd1.leo#@button test2
```

**outline filename only (only work in currently opened outline, but it should open other if file in same folder)**
```
unl:gnx://testmd1.leo#felix.20231224140524.1

unl://testmd1.leo#@button test2
```

**no outline (do not work at all, should work if in current outline)** 
```
unl://#@button test2

unl:gnx://#felix.20231224140524.1
```

I hope this helps to narrow down the remaining ctrl-click issues on some particular types of UNLs 😄

**Comment by tbpassin** [comment]
> outline full path (work well even with backslashes and get opened if not opened)

As I understand it, the new version of the old, path-based UNLs isn't supposed to contain the full path any more but only the filename.  So there is now an algorithm for tracking down the actual full path.  You'd have to ask Edward the rationale for this. but I contributed to expanding the algorithm for where to look. (I'm not sure if this version has been merged or not; I thought it was but I may be mistaken)

The algorithm goes something like this but I may not be remembering exactly right:
1. Check if it's the current outline;
2. Check if it's an open outline;
3. Check if it's in an `@data` setting that contains the full paths to various file names.
4. Check if it's in .leo
5. Check if it's a well-known file like LeoPyRef.leo.

If this expanded search algorithm has not been merged it it ought to be before any more changes are made to the regexes.  That's so we can make sure we've got a good collection of test cases.

I'm not sure if the same algorithm is supposed to be applied to the new gnx-style UNLs.

We are laboring under a handicap - there is no real spec for these UNLs and how they are supposed to work.  There is information in various docstrings, but it isn't comprehensive, and it needs to apply consistently for methods that create and use UNLs as well as for the regexes. So I expect to keep seeing bugs for some time.

**Comment by edreamleo** [comment]
On Thu, Dec 28, 2023 at 10:24 PM Thomas Passin ***@***.***>
wrote:

> outline full path (work well even with backslashes and get opened if not
> opened)
>
> As I understand it, the new version of the old, path-based UNLs isn't
> supposed to contain the full path any more but only the filename.
>
Full paths are always allowed. And now they work :-)

See the tests in TestGlobals.test_g_findAnyUnl, in the ekr-3736-unls
branch, that is, in PR #3737
<https://github.com/leo-editor/leo-editor/pull/3737>.

Edward

**Comment by tbpassin** [comment]
I believe this issue has been handled by the recent revisions to the UNL code, isn't that right?  If so, I think we can close it.

**Comment by edreamleo** [comment]
@tbpassin Thanks for the reminder, Thomas. Yes, we can close this.

**Comment by boltex** [comment]
The short version, where only the file name is given instead of the **full** path name, fail to open the file even if it's in the same folder as the currently opened file, or even fails to go to the unl even if it's already opened.

They only work in the currently opened document. (which renders the short file name format irrelevant) 

I'm going to reopen this issue. Please feel free to advise if i'm not understanding the expected behavior as described in my comment above at this link https://github.com/leo-editor/leo-editor/issues/3670#issuecomment-1871707526

Although some progress was made : the ones without any path at all now work in the same document: Here's a recap of the current state to compare with my original comment above:

**no outline at all _do work now_! (if in current outline, as expected)** 
```
unl://#@button test2

unl:gnx://#felix.20231224140524.1
```

**outline full path (work well, even with backslashes, and get opened if not opened)**
```
unl:gnx://C:\prog\testmarkdown\testmd1.leo#felix.20231224140524.1
unl:gnx://C:/prog/testmarkdown/testmd1.leo#felix.20231224140524.1

unl://C:/prog/testmarkdown/testmd1.leo#@button test2
unl://C:\prog\testmarkdown\testmd1.leo#@button test2
```

**outline filename only still only work in current outline (it should open other if file in same folder, as if full path.)**
```
unl:gnx://testmd1.leo#felix.20231224140524.1

unl://testmd1.leo#@button test2
```

**Comment by edreamleo** [comment]
@boltex I agree with reopening this issue. Thanks for the new explanation.

**Comment by edreamleo** [comment]
@tbpassin I have removed the Delegated label. I'll attempt a fix myself.

**Comment by tbpassin** [comment]
Hmmm, I thought I tested for opening other outlines in the same folder.  Details are hazy by now.  If we're re-opening this issue, let's remember to make sure that paths specified in that @data setting (I forget its name) will still get opened.

**Comment by tbpassin** [comment]
That `@data` setting is `@data unl-path-prefixes`.  The method most likely of interest is either `openUNLFile()` or `findAnyUnl()`.

**Comment by edreamleo** [comment]
Fixed via PR #3737 (Leo 3.7.7).

**Comment by boltex** [comment]
@edreamleo Why is this issue closed again? This something change somewhere to fix the issue? I don't get it... :/
